Support react 16 (quick way)

ngryman opened the pull request:
This PR takes the fastest path to support React 16 with one breaking change: the `domNode` passed to the component is already mutated so referring to its `textContent` will not work anymore.

I've skipped the associated test for now and wait for opinions on how to handle this.

cc @jdeal @vitorbal 
ref #10

ngryman committed:
support react 16

jdeal commented:
@ngryman Can you describe why the breaking change occurs?

jdeal commented:
It seems like this just keeps using the unstable API and so avoids the breaking change?

ngryman committed:
update node version on travis

ngryman committed:
remove react warning in tests

ngryman committed:
unskip the failing test

ngryman commented:
@jdeal Fixed the CI to highlight the issue: https://travis-ci.org/zapier/react-element-portal/builds/290528365#L1499

jdeal commented:
I see now @ngryman. That's definitely a breaking change. I think we'll want to allow passing in a callback instead. So something like:

```
const Greeting = ({textContent}) => (<div>Hello {textContent}</div>);
const mapDomNodeToProps = (domNode) => ({
  textContent: domNode.textContent
});
render(
  <div>
    <ElementPortal selector="li.greeting" view={Greeting} mapDomNodeToProps={mapDomNodeToProps}/>
  </div>,
  document.getElementById(appId)
);  
```

That way we can call the callback _before_ we render, similar to how we grab props from `dataset`. Effectively, `dataset` will become the preferred way to pass data, and `mapDomNodeToProps` will be a way to customize that.

ngryman commented:
@jdeal Yeah I've thought about that and it's the better compromise I think. If we're doing so, why not having only this callback and deprecate the `data` attribute?

jdeal commented:
@ngryman I'm _okay_ with that, but I do feel like the `data` attribute is pretty handy here and feels like a good recommended approach for passing data. Of course, it's also wasted cycles if you have a lot of attributes you don't use. But I don't know if that's a _huge_ concern here. I'm kinda torn. Maybe we add `mapDomNodeToProps` and hold off on the `data` decision for now? We _could_ export a helper I guess for that purpose too. I wouldn't want to force the `dataset` boiler-plate out to userland.

ngryman commented:
`mapDomNodeToProps` was what I was thinking, it gives lots of flexibility to the user while removing complexity our side.

> I'm okay with that, but I do feel like the data attribute is pretty handy here and feels like a good recommended approach for passing data.

With the introduction of `mapDomNodeToProps`, the user now has a generic way of extracting data from the pre-existing target node and pass it to the transported component.
IMHO, keeping the `data` attribute would be redundant and opinionated here. If the user needs to parse data attributes, then he could use a utility package for that (https://github.com/rafaelrinaldi/data-attributes):

```js
import dataAttributes from 'data-attributes'

const mapDomNodeToProps = (node) => ({
  data: dataAttributes(node)
})

<ElementPortal selector=".target" view={Component} mapDomNodeToProps={mapDomNodeToProps} />
```

ngryman committed:
introduce mapDomNodeToProps

Breaking change: The new way of passing data from the target node to the transported component is implemented via the mapDomNodeToProps function that takes the target node as input and binds the output object to the transported component.

ngryman committed:
remove todo comment

ngryman committed:
bump dependencies and removed unused ones

ngryman committed:
add package-lock.json

jdeal commented:
@ngryman I'm cool with that change, but can you replace the current `data` example in the README with your snippet using the `data-attributes` module? And we should probably publish this as 1.0.0?

ngryman committed:
add a module field pointing to es6 code

Let bundlers use the source instead of the pre-bundled version. Also very helpful for ad-hoc testing.

ngryman committed:
mention mapDomNodeToProps

ngryman commented:
I've updated the docs, please correct/enhance my English :)

While patching `zapier/zapier` to check things work correctly (and it works btw 🎉), I think we could take the chance to change the API a little bit for something a little more concise. Nothing really fancy, just some tweaks that I'm gonna put in a PR tomorrow.

Meanwhile, if you think everything is ok for the React 16 migration, please tell me so I can reference it in `zapier/zapier` and ask for a review there.

jdeal commented on README.md:
```diff
@@ -86,31 +89,50 @@ The `shouldReset` prop can be used to remove any classes and styles from the DOM
 
-One advantage of using the `view` prop to specify a component is that any `data-` attributes from the DOM node the portal is rendering to will be passed along to our component as a `data` prop.
-For example, if the DOM node we are rendering to looks like this:
+One advantage of using the `view` prop to the ability to derive properties from the original DOM node to the the component.
```
Probably want something like:

> One advantage of using the `view` prop is the ability to derive properties from the original DOM node and pass them to the the component.

jdeal commented on README.md:
````diff
@@ -86,31 +89,50 @@ The `shouldReset` prop can be used to remove any classes and styles from the DOM
+)
+```
+
+By using the `mapDomNodeToProps` you can easily pass this data like so:
````
Small tweak here:

> By using the `mapDomNodeToProps` prop, you can easily pass this data like so:

jdeal reviewed (approved):
@ngryman A couple small wording tweaks, but otherwise we're good!

ngryman committed:
docs tweaks
